src/main/cpp/src/api/genomicsdb.cc

```diff
@@ -725,3 +800,671 @@ void GenomicsDBVariantCallProcessor::process(const std::string& sample_name,
+       }
+      return;
+    }
+  }
+
+  int16_t ploidy = gt_vec.size();
+
+  if(!state) {
+    sample_map_initialized = true; // possible to skip first pass, sample map will be populated from callset
+
+    sample_map.insert(std::make_pair(coords[0], std::make_pair(-1, sample_name)));
+    last_coord = coords[1];
+    return;
+  }
+
+  if(ploidy != 2 && (make_bed || make_tped)) { 
+    logger.error("The tped/bed formats do not support ploidies other than 2.");
+    make_bed = 0;
+    make_tped = 0;
+    if(make_bgen) {
+      logger.info("Continuing bgen generation");
+    } 
+  }
+
+  if(state == 1) {
+    std::string rsid;
+    std::string rsid_row;
+    if(id_string.size()) {
+      rsid = id_string;
+    }
+    else {
+      rsid = genomic_interval.contig_name + ":" + std::to_string(genomic_interval.interval.first);
+    }
+    rsid_row = genomic_interval.contig_name + " " + rsid + " 0 " + std::to_string(genomic_interval.interval.first);
+
+    int sind = sample_map[coords[0]].first;
+
+    // backfill if needed
+    bool backfill = coords[1] - last_coord && last_coord != -1;
+    int add_to_prev = backfill ? sample_map.size() - (last_sample + 1) : 0;
+    int add_to_current = backfill ? sind : sind - (last_sample + 1);
+
+    if(last_coord != coords[1]) { // first in variant
+      num_variants++;
+
+      for(int i = 0; i < add_to_prev; i++) { // backfill samples missing from previous variant
+        if(make_tped) {
+          tped_file << " 0 0";
+        }
+        if(make_bed) {
+          write_to_bed(1);
+        }
+        if(make_bgen) {
+          // BGEN: backfill probability data
+          bgen_empty_cell(2, last_alleles, phased);
+        }
+      }
+      if(make_bed) {
+        flush_to_bed();
+      }
+
+      if(make_tped) {
+        if(last_sample != -1) { // first line should not have newline
+          tped_file << std::endl;
+        }
+        tped_file << rsid_row;
+      }
+
+      if(make_bgen) {
+        bgen_variant_data_block(rsid, genomic_interval, vec, phased);
+      }
+    }
+    else {
+      samples_in_column++;
+    }
+
+    for(int i = 0; i < add_to_current; i++) { // backfill samples missing from current variant
+      if(make_bed) {
+        write_to_bed(1);
+      }
+
+      if(make_tped) {
+        tped_file << " 0 0";
+      }
+
+      if(make_bgen) {
+        // BGEN: backfill probability data
+        bgen_empty_cell(2, vec.size(), phased);
+      }
+    }
+
+    // safe to update now that backfilling is over
+    last_alleles = vec.size();
+
+    char gt_code;
+
+    if(!(gt_vec[0] || gt_vec[1])) { // homozygous for first allele
+      gt_code = 0;
+    }
+    else if(gt_vec[0] + gt_vec[1] == 1) { // heterozygous
+      gt_code = 2;
+    }
+    else if(gt_vec[0] == 1 && gt_vec[1] == 1) { // homozygous for second allele
+      gt_code = 3;
+    }
+    else {
+      gt_code = 1;
```

Comment:
This is maybe beyond the scope of this PR but...since bed only really supports two alleles (right?) and if i'm remembering right, its just a primary and secondary allele - no real notion of reference or alternate.

With this implementation, we will spit out a `1` (or missing genotype) in cases where at least one of the alleles represented in the genotype are not the reference or the first alternate allele. The "correct" thing to do would be to find the two alleles represented by the genotype, and spit those out instead of reference+first alternate. And then, missing genotype (`1`) would really just be missing. 

I'm fine with not fixing this here...but can we add a FIXME, and maybe also an issue?

Reply:
Now uses the alleles from GT rather than REF and ALT, as a result will only ever use codes 0 and 2.